pyat/at/lattice/lattice_object.py

```diff
@@ -312,21 +302,136 @@ def _addition_filter(self, elems: Iterable[Element]):
+                                     If :py:obj:`True` deep copies of each
+                                     element of elems are used
+        """
         if hasattr(self, '_energy'):
             # When unpickling a Lattice, extend is called before the lattice
             # is initialized. So skip this.
-            elems = self._addition_filter(elems)
+            elems = self._addition_filter(elems, copy_elements=copy_elements)
         super().extend(elems)
 
-    def append(self, elem: Element):
-        self.extend([elem])
+    def append(self, elem: Element, copy_elements=False):
+        r"""This method overwrites the inherited method
+            :py:meth:`list.append()`,
+            it behavior is changed, it accepts only AT lattice elements
+            :py:obj:`Element` as input argument.
+
+            Equivalents syntaxes:
+            >>> ring.append(elem)
+            >>> ring += [elem]
+
+            Parameters:
+                elem (Element): AT element to be appended to the lattice
+
+            Keyword Arguments:
+                copy_elements(bool): Default :py:obj:`True`.
+                                     If :py:obj:`True` a deep copy of elem
+                                     is used
+        """
+        self.extend([elem], copy_elements=copy_elements)
+
+    def repeat(self, n: int, copy_elements=True):
+        r"""This method allows to repeat the lattice `n` times.
+            If `n` does not divide `ring.periodicity`, the new ring
+            periodicity is set to 1, otherwise  it is et to
+            `ring.periodicity /= n`.
+
+            Equivalents syntaxes:
+            >>> newring = ring.repeat(n)
+            >>> newring = ring * n
+
+            Parameters:
+                n (int): number of repetition
+
+            Keyword Arguments:
+                copy_elements(bool): Default :py:obj:`True`.
+                            If :py:obj:`True` deepcopies of the
+                            lattice are used for the repetition
+
+            Returns:
+                newring (Lattice): the new repeated lattice
+        """
+        def copy_fun(elem, copy):
+            if copy:
+                return elem.deepcopy()
+            else:
+                return elem
+
+        periodicity = self.periodicity
+        if n != 0 and periodicity > 1:
+            nbp = periodicity / n
+            periodicity = int(round(nbp))
+            if abs(periodicity - nbp) > _TWO_PI_ERROR:
+                warn(AtWarning('Non-integer number of cells: {}/{}. Periodi'
+                               'city set to 1'.format(self.periodicity, n)))
+                periodicity = 1
+        elems = (copy_fun(el, copy_elements) for _ in range(n) for el in self)
+        return Lattice(elem_generator, elems, iterator=self.attrs_filter,
+                       periodicity=periodicity)
+
+    def concatenate(self, *lattices: Tuple[Iterable[Element], ...],
+                    copy_elements=False, copy=True):
```

Comment:
In all the examples where the `copy` keyword is used (`enable_6d`, `set_cavity`, `set_value_refpts`, `add_beamloading`), its default if `False`. It's not absolutely necessary to have the same default everywhere, but for me it would be easier to remember… 

Anyway, whatever the default value is, I'd push to explicitly specify `copy` in all internal calls to concatenate (within `lattice_object.py`). It's clearer!